libr/io/p/io_gdb.c

```diff
@@ -229,6 +235,157 @@ static char *__system(RIO *io, RIODesc *fd, const char *cmd) {
 			                     " current/specified pid\n");
 		return NULL;
 	}
+	if (cmd[0] == 'd') {
+		switch (cmd[1]) {
+		case 'c':
+			switch (cmd[2]) {
+			case '\0': // "dc"
+				gdbr_continue (desc, desc->pid, -1, 0);
+				if (desc->stop_reason.is_valid && desc->stop_reason.thread.present) {
+					desc->tid = desc->stop_reason.thread.tid;
+				}
+				return NULL;
+			case 'b': // "dcb"
+				if (!desc->stub_features.ReverseContinue) {
+					eprintf ("Continue backwards is not supported in this gdbserver implementation\n");
+					return NULL;
+				}
+				gdbr_lock_enter (desc);
+				if (send_msg (desc, "bc") >= 0) {
+					(void)read_packet (desc);
+					desc->data[desc->data_len] = '\0';
+					if (!desc->no_ack) {
+						eprintf ("[waiting for ack]\n");
+					} else {
+						handle_stop_reason (desc);
+						if (desc->stop_reason.is_valid == false) {
+							eprintf ("Thread (%d) stopped for an invalid reason: %d\n",
+								desc->stop_reason.thread.tid, desc->stop_reason.reason);
+						}
+					}
+					gdbr_invalidate_reg_cache ();
+				}
+				gdbr_lock_leave (desc);
+				return NULL;
+			case 'k': // "dck"
+				if (cmd[3] != '\0') {
+					int sig = r_num_math (NULL, cmd + 3);
+					gdbr_continue (desc, desc->pid, -1, sig);
+					if (desc->stop_reason.is_valid && desc->stop_reason.thread.present) {
+						desc->tid = desc->stop_reason.thread.tid;
+					}
+				}
+				return NULL;
+			default:
+				break;
+			}
+			break;
+		case 'm': // "dm"
+		{
+			char path[128];
+			ut8 *buf;
+			int ret;
+			// TODO refer debug_gdb.c:r_debug_gdb_map_get
+			ut64 buflen = 16384;
+			snprintf (path, sizeof (path) - 1, "/proc/%d/maps", desc->pid);
+
+#ifdef _MSC_VER
+#define GDB_FILE_OPEN_MODE (_S_IREAD | _S_IWRITE)
+#else
+#define GDB_FILE_OPEN_MODE (S_IRUSR | S_IWUSR | S_IXUSR)
+#endif
+
+			if (gdbr_open_file (desc, path, O_RDONLY, GDB_FILE_OPEN_MODE) < 0) {
+				return NULL;
+			}
+			if (!(buf = malloc (buflen))) {
+				gdbr_close_file (desc);
+				return NULL;
+			}
+			if ((ret = gdbr_read_file (desc, buf, buflen - 1)) <= 0) {
+				gdbr_close_file (desc);
+				free (buf);
+				return NULL;
+			}
+			buf[ret] = '\0';
+
+			char *ptr = (char *)buf;
+			RStrBuf *obuf = r_strbuf_new (NULL);
+			ut64 map_start, map_end;
+			char perm[32];
```

Comment:
Despite being enough i dont like fixed size buffers . Maybe worth to restrict the sscanf read size in the format string